keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1076,361 @@ the remaining requests in that queue start getting faster service.  In
+
+Similarly as in generic case of long-running requests, predicting this cost
+up front is impossible. Fortunately, in this case we have a full control over
+those events, because those are effectively a result of the mutating requests
+that out mechanism is explicitly admitting.
+So instead of associating the cost of sending watch events to the corresponding
+WATCH request, we will reverse the situation and associate the cost of sending
+watch events to the mutating request that triggering them.
+
+In other words, we will be throttling the write requests to ensure that
+apiserver would be able to keep up with the watch traffic instead of throttling
+the watchers themselves.  The main reason for that is that throttling watchers
+themselves isn't really effective: we either need to send all objects to them
+anyway or in case we close them, they will try to resume from last received
+event anyway.  Which means we don't get anything on throttling them.
+
+##### Estimating cost of the request
+
+Let's start with an assumption that sending every watch event is equally
+expensive.  We will discuss how to generalize it below.
+
+With the above assumption, a cost of a mutating request associated with
+sending watch events triggerred by it is proportional to the number of
+watchers that has to process that event.   So let's describe how we can
+estimate this number.
+
+Obviously, we can't afford going over all watches to compute that - we need
+to keep this information already precomputed. What if we would simply
+store an in-memory map from (resource type, namespace, name) tuple into
+number of opened watches that are interested in a given event.  The size of
+that map won't be larger then the total number of watches, so that is
+acceptable.
+Note that each watch can also specify label and field selectors.  However,
+in most of cases a particular object has to be processed for them anyway
+to check if the selectors are satisfied.  So we can ignore those selectors
+as the object contributes to the cost (even if it will not be send as not
+satisfying the selector).
+The only exception to this is caused by a few predefined selectors that
+kube-apiserver is optimized for (this includes pods from a given node and
+nodes/secrets/configmaps specifying metadata.name field selector).  Given
+their simplicity, we can extend our mapping to handle those too.
+
+Having such in-memory map, we can quickly estimate the cost of a request.
+It's not as simple as taking a single map item as it requires adding watches
+for the whole namespace and all objects of a given type.  But it be done in
+O(1) map accesses.
+Keeping such map up-to-date is also easy - whenever a new watch start we
+increment a corresponding entry, when it ends we decrement it.
+
+##### Multiple apiservers
+
+All the above works well in the case of single kube-apiserver.  But if there
+are N kube-apiservers, there is no guarantee that the number of watches are
+evenly distributed across them.
```

Comment:
This is true of all sorts of requests.

Reply:
Agree - but others don't matter much.
If I have 3 watches for object X and the second apiserver has 3000, the change of object X will be more-or-less 1000 more expensive for the second apiserver. So I may admit 1000x changes than the second apiserver could keep up with, but I would keep up with it.